XMonad/Util/PureX.hs

```diff
@@ -271,6 +271,14 @@ shift tag = withFocii $ \ctag fw ->
     mfw' <- peek
     return (Any $ Just fw /= mfw')
 
+-- | A refresh tracking version of @W.shiftWin@
+shiftWin :: XLike m => WorkspaceId -> Window -> m Any
+shiftWin tag w = do
+  ctag <- curTag
+  when' (tag /= ctag) $ do
+    modifyWindowSet' (W.shiftWin tag w)
+    return $ Any True
```

Comment:
`W.shiftWin` can end up being a no-op if `tag` refers to a non-existent workspace or if `w` refers to a non-existent window. Not sure how big a deal this is, but the other functions in this module handle that correctly.

Reply:
Latest commits fix this, and a separate logic error (the old implementation assumed that the window `w` was on the current workspace)